A stable, safe implementation of "Struct Target Features"

DJMcNab opened the pull request:
The core contributions of this PR are:

- A trait which a (should be zero-sized) struct can implement, which indicates that it is a type-level proof that a set of target features are enabled.
- The `trampoline` macro, which validates a `#[target_feature(enable = "xxx")]` string against values of one or more of these, ensuring at compile time that a call to a `#[target_feature]` function will be safe; and then calling it.
- A corresponding struct for each target feature on x86[-64], which are code generated.

The state of this feature is:

- It is *not* used for implementing the Fearless SIMD crate.
- ~~The x86-64-v{1,2,3,4} level implementations do not exist/are extremely incomplete.~~
- ~~Some docs are missing (these are however not the most critical docs, it's only docs on the groupings of x86 features).~~
- It does not have support for aarch64 in the architecture levels. This is not hard, it's just data wrangling.

There is also an open licensing question, around the docs taken from the Rust reference. My preference would be to copy https://github.com/rust-lang/reference/blob/1d930e1d5a27e114b4d22a50b0b6cd3771b92e31/LICENSE-MIT#L1 into our LICENSE-MIT, which avoids having to make a decision about copyright-ability here.

My proposed next steps are:

- Discuss this at Renderer Office Hours tomorrow: Done
- If we decide this is a direction we want to follow, clean up and land this PR.
   - Bump our MSRV to 1.89 to take https://github.com/rust-lang/rust/issues/111137 into account: Done
   - Add final docs: Done
   - Validate that things are up-to-date in CI: Done
   - True x86-64-v{1,2,3,4} level support: Done
- Follow-up with:
   - aarch64 support
   - Automatic selection/an enum of x86-64 levels
   - Using it in the implementation of Fearless SIMD itself

For review:

- You can mostly ignore the contents of `fearless_simd_core/x86/xxx/xxx.rs`, as these are entirely automatically generated. The exception is `fearless_simd_core/x86/xxx/mod.rs`, which are hand-written, but don't have any logic.

Discussed on Zulip: [#simd > Removing &#96;safe-wrappers&#96;](https://xi.zulipchat.com/#narrow/channel/514230-simd/topic/Removing.20.60safe-wrappers.60/with/545377857)

DJMcNab committed:
Save the version before running the x86 generator

DJMcNab committed:
Remove old v1 items

DJMcNab committed:
Improve the generator

DJMcNab committed:
Minor fixups in the generator

Also removes unused additional impl support

DJMcNab committed:
Add the generated x86 code

DJMcNab committed:
Fixup some docs

DJMcNab committed:
Save some generator changes, including adding x86_v{1,2,3,4}

DJMcNab committed:
Fixup the generator and mostly finalize levels

DJMcNab committed:
Re-run update

DJMcNab committed:
Bump MSRV to allow avx512 support

DJMcNab committed:
Misc cleanups to get ready to launch

DJMcNab committed:
Handle sse4a and tbm consistently

DJMcNab committed:
Add the final missing docs

DJMcNab committed:
Add copyright headers

DJMcNab committed:
Add a CI check for the new generator

DJMcNab committed:
Fixup docs on `vectorize`

taj-p commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+//! other `#[target_feature]` functions.
+//! As such, once you have used the [`trampoline!`] macro, you can call any intrinsic in [`core::arch`].
+//!
+//! This crate also has modules which contain tokens for each Rust target features.
```
```suggestion
//! This crate also has modules which contain tokens for each Rust target feature.
```

taj-p commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+//!
+//! # Examples
+//!
+//! At the time of writing, it is not possible to turn scalar values into <abbr title="Single InstructionMultiple Data">SIMD</abbr>
```
```suggestion
//! At the time of writing, it is not possible to turn scalar values into <abbr title="Single Instruction Multiple Data">SIMD</abbr>
```

ajakubowicz-canva replied:
```suggestion
//! At the time of writing, it is not possible to turn scalar values into <abbr title="Single Instruction Multiple Data">SIMD</abbr>
```

taj-p commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+//! vector types safely using only the standard library.
+//! These examples use [bytemuck](https://crates.io/crates/bytemuck) for this.
+//!
+//! <!-- TODO -->
```
Is this TODO deliberately left for completing in a later PR?

DJMcNab replied:
Yeah, that was my approximate intention. I'll plan to follow up with that very soon after we land this, but I don't want to potentially block this entire PR on getting those reviewed now.

taj-p commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+//!
+//! # Crate Feature Flags
+//!
+//! <!-- TODO -->
```
Is this a deliberately left TODO?

taj-p commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+#[cfg(feature = "std")]
+extern crate std;
+
+/// Token that a set of target feature is available.
```
```suggestion
/// Token that indicates a set of target features is available.
```

taj-p commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+/// That is, to enable the target features in `FEATURES`, you *must* have a value
+/// of this type.
+///
+/// Values which implement this trait are used in the second argument to [`trampoline!`],
```
```suggestion
/// Values which implement this trait are used in the first argument to [`trampoline!`],
```

I thought it was passed to the first argument of `trampoline!`

ajakubowicz-canva commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+//! vector types safely using only the standard library.
+//! These examples use [bytemuck](https://crates.io/crates/bytemuck) for this.
+//!
+//! <!-- TODO -->
```
Just to clarify is this awaiting a code example? May be helpful to document what the TODO is awaiting.

taj-p reviewed (commented):
Soundness makes sense to me! Let me know if these TODOs are deliberate, once you've solved the question of whether to split out `x86` into `x86` and `x86_64`, and I'm happy to approve

ajakubowicz-canva commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+///
+/// This is effectively a stable implementation of the "Struct Target Features" Rust feature,
+/// which at the time of writing is neither in stable or nightly Rust.
+/// This macro can be used to make SIMD dispatch safe in addition to make explicit SIMD, both safely.
```
Nit: This line is a bit awkward to read.

ajakubowicz-canva commented on fearless_simd_core/src/trampoline.rs:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    #[test]
+    fn empty_feature() {
```
`expect_success(",,,,,,", [&[""]]);` – list of empty target features.

ajakubowicz-canva commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+//! This abstraction is designed to be combined with target features 1.1, the recent update
+//! in the Rust compiler to allow calling `#[target_feature]` functions safely from within
+//! other `#[target_feature]` functions.
+//! As such, once you have used the [`trampoline!`] macro, you can call any intrinsic in [`core::arch`].
```
I think the "glamour shot" described in comment https://github.com/linebender/fearless_simd/pull/108#issuecomment-3407267104 makes perfect sense to me now! I really like this PR. Apologies that it's taking me a while to get through. On Monday I plan to go through the code that's outside fearless_simd_core.

DJMcNab commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+///
+/// # Reference
+///
+/// These reference examples presume that you have (values in brackets are the "variables"):
```
So, this is the intro to the below list. But you're right that it reads strangely, I'll reword!

DJMcNab commented on fearless_simd_core/src/lib.rs:
```diff
@@ -0,0 +1,339 @@
+/// let a = [1., 2., 3., 4.];
+/// let Some(token) = token else { return scalar_fallback(a) };
+///
+/// trampoline!(Token = token => "f1,f2", uses_simd(a: [f32; 4]) -> [f32; 4])
```
Yeah, exactly. This is due to our dependency of the target features 1.1 Rust feature, which moves that safety check into the Rust compiler.

And yes, you can use multiple tokens exactly as you describe. I'll see about adding some docs for that.

DJMcNab committed:
Address various review feedback and other cleanups

Rename `trampoline.rs` to `support.rs`
The old name conflicted with the name of the macro, leading to it being
harder to find the docs of the macro itself.

Remove unneeded reference

Remove entire note on 128 bytes being too small

The point it was making was:
- Fairly hard to explain
- Not necessarily true

Add a few more test cases

Co-authored-by: Taj Pereira <taj@canva.com>

DJMcNab committed:
Clean up the stuff about licensing

DJMcNab committed:
Add SSE4a and TBM with MSRV 1.91
